catwalk/models/gpt.py

```diff
@@ -1,34 +1,100 @@
+
+from torch import log_softmax
+from torch.nn.utils.rnn import pad_sequence
+from transformers import AutoModelForCausalLM, AutoTokenizer
 
 
 @Model.register("catwalk::gpt")
 class GPTModel(Model):
     def __init__(self, pretrained_model_name_or_path: str):
         self.pretrained_model_name_or_path = pretrained_model_name_or_path
+        
+    def _convert_instances(self, instances: Sequence[Dict[str, Any]], instance_format, task) -> MappedSequence:
+        return MappedSequence(lambda instance: task.convert_instance(instance, instance_format), instances)
 
     def predict(  # type: ignore
+        self,
+        task: Task,
+        instances: Sequence[Dict[str, Any]],
+        *,
+        batch_size: int = 32,
+    ) -> Iterator[Dict[str, Any]]:
+        if task.has_instance_conversion(InstanceFormat.HF_QA):
+            return self._predict_qa(task, instances, batch_size=batch_size)
+
+        return self._predict_perplexity(task, instances, batch_size=batch_size)
+
+    def _predict_qa(
+            self,
+            task: Task,
+            instances: Sequence[Dict[str, Any]],
+            *,
+            batch_size: int = 32
+    ) -> Iterator[Dict[str, Any]]:
+        device = resolve_device()
+        model = cached_transformers.get(
+            AutoModelForCausalLM, self.pretrained_model_name_or_path, False).eval().to(device)
+        tokenizer = cached_transformers.get_tokenizer(
+            AutoTokenizer, self.pretrained_model_name_or_path)
+        tokenizer.pad_token = tokenizer.eos_token
+        tokenizer.padding_side = 'left'
+        
+        converted_instances = self._convert_instances(instances, InstanceFormat.HF_QA, task)
+
+        def format_instance(instance: HFQAInstance) -> Tuple[str, str]:
+            # TODO: Use promptsource to add more prompt options?
+            return instance.context, f"\nQuestion:{instance.question}\nAnswer:"
+
+        converted_instances = Tqdm.tqdm(converted_instances, desc="Processing instances")
+        for batch in more_itertools.chunked(converted_instances, batch_size):
+            formatted_batch = [format_instance(instance) for instance in batch]
+            encodings = tokenizer(formatted_batch,
+                                  padding="longest",
+                                  truncation="only_first",
+                                  max_length=model.config.n_positions - model.config.task_specific_params['text-generation']['max_length'],
+                                  return_tensors="pt")
+            
+            with torch.inference_mode():
+                outputs = model.generate(input_ids=torch.stack(encodings["input_ids"]).to(device),
+                                         attention_mask=torch.stack(
+                                             encodings["attention_mask"]).to(device),
+                                         max_new_tokens=model.config.task_specific_params['text-generation']['max_length'])
```

Comment:
The question remains how padding works here. As @jagnusson points out, regardless of which side you put the padding on, you have to do something with position ids. Or maybe `model.generate()` takes care of that internally? We should know though. In Ian's case, it made a significant difference to performance.

Reply:
I looked into this and I believe that this is taken care on inside the generate method using the attention mask as seen [here](https://github.com/huggingface/transformers/blob/21f6f58721dd9154357576be6de54eefef1f1818/src/transformers/models/gpt2/modeling_gpt2.py#L1014). This is also confirmed in this [issue](https://github.com/huggingface/transformers/issues/18478) in a code example from a Huggingface engineer.